src/main/java/de/braintags/vertx/util/json/JsonDiff.java

```diff
@@ -481,6 +483,142 @@ private static void squashArrayDiff(final ArrayNode base, final ArrayNode diff,
     base.addAll(Arrays.asList(newValues));
   }
 
+  /**
+   * Retains the diff tree specified by retainedFields
+   *
+   * @param base
+   *          is not modified
+   * @param retainedFields
+   *          is not modified
+   * @param diff
+   *          MAY BE MODIFIED!
+   * @return
+   *         the modified JsonNode
+   */
+  public static JsonNode retainDiffTree(final JsonNode base, final JsonNode retainedFields, final JsonNode diff) {
+    return internalRetainDiffTree(base, retainedFields, diff, Json.mapper.getNodeFactory(), false);
+  }
+
+  private static JsonNode internalRetainDiffTree(final JsonNode base, final JsonNode retainedFields,
+      final JsonNode diff,
+      final JsonNodeFactory nodeFactory, final boolean arrayMapsConverted) {
+    JsonNodeType nodeType = base.getNodeType();
+    if (nodeType == null) {
+      if (base instanceof ArrayMapNode && diff instanceof ArrayMapNode) {
+        ArrayMapNode arrayBase = (ArrayMapNode) base;
+        ArrayMapNode retainedArrayDiff = (ArrayMapNode) ArrayMapNode.deepConvertNode(retainedFields);
+        ArrayMapNode arrayDiff = (ArrayMapNode) ArrayMapNode.deepConvertNode(diff);
+        retainDiffTreeArrayMapDiff(arrayBase, retainedArrayDiff, arrayDiff, nodeFactory, true);
+        return arrayDiff;
+      } else {
+        throw new IllegalArgumentException("nodeType is null and base and/or diff is not ArrayMapNode");
+      }
+    }
+
+    switch (nodeType) {
+      case OBJECT:
+        if (!arrayMapsConverted && ArrayMapNode.isArrayMapNode(base)) {
+          ArrayMapNode arrayBase = (ArrayMapNode) ArrayMapNode.deepConvertNode(base);
+          ArrayMapNode retainedArrayDiff = (ArrayMapNode) ArrayMapNode.deepConvertNode(retainedFields);
+          ArrayMapNode arrayDiff = (ArrayMapNode) ArrayMapNode.deepConvertNode(diff);
+          retainDiffTreeArrayMapDiff(arrayBase, retainedArrayDiff, arrayDiff, nodeFactory, true);
+          return ArrayMapNode.toRegularNode(arrayDiff, nodeFactory);
+        } else if (diff.isObject()) {
+          retainDiffTreeObjectDiff((ObjectNode) base, (ObjectNode) retainedFields, (ObjectNode) diff, nodeFactory,
+              arrayMapsConverted);
+          return diff;
+        } else {
+          return diff.deepCopy();
+        }
+      case ARRAY:
+        retainDiffTreeArrayDiff((ArrayNode) base, (ArrayNode) retainedFields, (ArrayNode) diff, nodeFactory,
```

Comment:
if you special case arrays and call a function that just throws an exception, maybe add an test that verifies this ? Otherwise regressions may happen an arrays suddenly get handled again

Reply:
done